internal/server/singleprocess/service_auth_oidc.go

```diff
@@ -0,0 +1,305 @@
+		return nil, status.Errorf(codes.FailedPrecondition,
+			"auth method is not OIDC")
+	}
+
+	// We need our server config.
+	sc, err := s.state.ServerConfigGet()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get our OIDC provider
+	provider, err := s.oidcCache.Get(ctx, am, sc)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a minimal request to get the auth URL
+	oidcReqOpts := []oidc.Option{
+		oidc.WithNonce(req.Nonce),
+	}
+	if v := amMethod.Oidc.Scopes; len(v) > 0 {
+		oidcReqOpts = append(oidcReqOpts, oidc.WithScopes(v...))
+	}
+	oidcReq, err := oidc.NewRequest(
+		oidcReqExpiry,
+		req.RedirectUri,
+		oidcReqOpts...,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the auth URL
+	url, err := provider.AuthURL(ctx, oidcReq)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.GetOIDCAuthURLResponse{
+		Url: url,
+	}, nil
+}
+
+func (s *service) CompleteOIDCAuth(
+	ctx context.Context,
+	req *pb.CompleteOIDCAuthRequest,
+) (*pb.CompleteOIDCAuthResponse, error) {
+	log := hclog.FromContext(ctx)
+
+	if err := serverptypes.ValidateCompleteOIDCAuthRequest(req); err != nil {
+		return nil, err
+	}
+
+	// Get the auth method
+	am, err := s.state.AuthMethodGet(req.AuthMethod)
+	if err != nil {
+		return nil, err
+	}
+
+	// The auth method must be OIDC
+	amMethod, ok := am.Method.(*pb.AuthMethod_Oidc)
+	if !ok {
+		return nil, status.Errorf(codes.FailedPrecondition,
+			"auth method is not OIDC")
+	}
+
+	// We need our server config.
+	sc, err := s.state.ServerConfigGet()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get our OIDC provider
+	provider, err := s.oidcCache.Get(ctx, am, sc)
+	if err != nil {
+		return nil, err
+	}
+
+	// We need this discovery info later
+	discovery, err := provider.DiscoveryInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a minimal request to get the auth URL
+	oidcReqOpts := []oidc.Option{
+		oidc.WithNonce(req.Nonce),
+		oidc.WithState(req.State),
+	}
+	if v := amMethod.Oidc.Scopes; len(v) > 0 {
+		oidcReqOpts = append(oidcReqOpts, oidc.WithScopes(v...))
+	}
+	if v := amMethod.Oidc.Auds; len(v) > 0 {
+		oidcReqOpts = append(oidcReqOpts, oidc.WithAudiences(v...))
+	}
+	oidcReq, err := oidc.NewRequest(
+		oidcReqExpiry,
+		req.RedirectUri,
+		oidcReqOpts...,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Exchange our code for our token
+	oidcToken, err := provider.Exchange(ctx, oidcReq, req.State, req.Code)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract the claims for this token. We create a struct here that
+	// extracts only the claims we actually care about.
+	var idClaimVals idClaims
+	if err := oidcToken.IDToken().Claims(&idClaimVals); err != nil {
+		return nil, err
+	}
+
+	// Valid OIDC providers should never behave this way.
+	if idClaimVals.Sub == "" {
+		return nil, status.Errorf(codes.Internal, "OIDC provider returned empty subscriber ID")
+	}
+
+	// Look up a user by sub.
+	user, err := s.oidcInitUser(log, discovery.Issuer, idClaimVals.Sub, idClaimVals.Email)
```

Comment:
Watch out! The `email` claim might contain an unverified email and thus allow impersonation!

https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

Some identity providers (like Keycloak) will allow you to change your own email address to what you want, but will set the `email_verified` claim to false until you confirm your email address. Failing to use the `email_verified` claim could lead to a security issue since the users will be able to impersonate someone else.

Reply:
Yes, we did discuss this... but I don't think I fully implemented this correctly.

Our plan was to only link a user by email if their email was verified. I even extracted the verified claim but didn't use it. I'll fix this.